File: 09_number_theory/1_modular_arithmetic/1_basic_modular_arithmetic/3_binary_exponentiation/UVA11029_Leading and Trailing.cpp
#include <cmath>
#include <iostream>
using namespace std;

long long modExp(long long base, long long exp, long long mod) {
  long long result = 1;
  base = base % mod;
  while (exp > 0) {
    if (exp % 2 == 1) {
      result = (result * base) % mod;
    }
    base = (base * base) % mod;
    exp = exp / 2;
  }
  return result;
}

long long calculateFirstThreeDigits(long long base, long long exp) {
  double logValue = exp * log10(base);  // log(n^k) = k * log10(n)
  double fractionalPart = logValue - floor(logValue);  // Get fractional part
  long long firstThree =
      (long long)(pow(10, fractionalPart) * 100);  // Get first 3 digits

  /*more precious for finding first three digit
  ld lg=(ld)m*log10((ld)n); //ld = long double
                lg -= floor(lg) - 2;
                lg = pow((ld)10,lg);
                cout<<floor(lg+1e-9)<<"..."<<l<<endl;
  */
  return firstThree;
}

int main() {
  int t;
  cin >> t;
  while (t--) {
    long long n, k;
    cin >> n >> k;
    long long lastThree = modExp(n, k, 1000);
    long long firstThree = calculateFirstThreeDigits(n, k);
    cout << firstThree << "..." << lastThree << endl;
  }
  return 0;
}

// 2nd type

#define ld long double

long long calculateFirstThreeDigits(long long n, long long m) {
  ld lg = (ld)m * log10((ld)n);
  lg -= floor(lg) - 2;
  ld firstDigits = pow((ld)10, lg);
  return (long long)firstDigits + 1e-9;
}

string formatLastThreeDigits(long long digits) {
  string result = to_string(digits);
  while (result.length() < 3) {
    result = "0" + result;  // Pad with leading zeros if necessary
  }
  return result;
}

int main() {
  int t;
  cin >> t;
  while (t--) {
    long long n, k;
    cin >> n >> k;
    long long lastThree = modExp(n, k, 1000);
    long long firstThree = calculateFirstThreeDigits(n, k);
    cout << firstThree << "..." << formatLastThreeDigits(lastThree) << endl;
  }
  return 0;
}

File: _site/09_number_theory/3_modular_arithmetic/3_binaryExpo_mulmod_aLike/UVA11029_Leading and Trailing .java
import java.math.BigInteger;
import java.util.Scanner;

//optimise solution
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            BigInteger n = sc.nextBigInteger();
            BigInteger k = sc.nextBigInteger();

            // Calculate last three digits using modular arithmetic
            BigInteger mod = BigInteger.valueOf(1000);
            BigInteger lastThree = n.modPow(k, mod);

            // Calculate first three digits
            double logValue = k.doubleValue() * Math.log10(n.doubleValue());
            double fractionalPart = logValue - Math.floor(logValue);
            int firstThree = (int) (Math.pow(10, fractionalPart) * 100);

            // Format the output
            System.out.printf("%03d...%03d\n", firstThree, lastThree.intValue());
        }
        sc.close();
    }
}

//using binary expo
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            BigInteger n = sc.nextBigInteger();
            BigInteger k = sc.nextBigInteger();

            // Calculate last three digits using modular arithmetic
            BigInteger mod = BigInteger.valueOf(1000);
            BigInteger lastThree = n.modPow(k, mod);

            // Calculate first three digits using binary exponentiation
            double base = n.doubleValue();
            double leadingDigits = 1.0;
            BigInteger exp = k;
            BigInteger two = BigInteger.valueOf(2); // Use BigInteger.valueOf(2)

            while (exp.compareTo(BigInteger.ZERO) > 0) {
                if (exp.mod(two).equals(BigInteger.ONE)) { // Check if exp is odd
                    leadingDigits *= base;
                    while (leadingDigits >= 1000) {
                        leadingDigits /= 10; // Truncate to leading digits
                    }
                }
                base *= base;
                while (base >= 1000) {
                    base /= 10; // Truncate to leading digits
                }
                exp = exp.divide(two); // Divide exp by 2
            }
            int firstThree = (int) leadingDigits;

            // Format the output
            System.out.printf("%03d...%03d\n", firstThree, lastThree.intValue());
        }
        sc.close();
    }
}
